sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,131 @@
+
+**March 2018**
+
+This document serves as a catalog of issues we've known/discovered with kubernetes services as of today, focusing on their scalability/performance. The purpose of the document is to make the information common knowledge for the community, so we can work together towards improving it. Listing them below in no particular order.
+
+
+## Performance of iptables when large no. of service ports exist
+
+#### Issue
+
+Iptables can be slow in the presence of a large no. of services. As the no. of service ports increases, the KUBE-SERVICES chain gets longer and since it’s also evaluated very frequently that can bring down the performance. Also, it seems that manipulating a chain with lots of rules (e.g. KUBE-SERVICES) takes noticeably longer than others.
+
+#### Relevant issues/links
+
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+
+#### Possible Solution(s)
+
+- Moving from iptables to IPVS may help here if that works stably. We have the IPVS alternative [implemented](https://github.com/kubernetes/kubernetes/pull/46580), but it still hasn’t gone to GA. Also we need to measure/confirm performance benefits from it.
+
+
+## Kube-proxy slow in updating iptables when large no. of backends exist
+
+#### Issue
+
+When we’re running services worth a total of 120k endpoints (backends) on our 5k-node scalability tests, we’re noticing that kube-proxy is timing out while trying to do iptables-restore due to failing to acquire lock over iptables.
+Note: Need to confirm how bad the situation is currently after some improvements that were made in 1.8.
+
+#### Relevant issues/links
+
+- [#48107](https://github.com/kubernetes/kubernetes/issues/48107)
+
+#### Possible Solution(s)
+
+- Give more time for kube-proxy to update iptables by increasing its batching interval for large clusters
+- Decide to support a smaller no. of backends for kubernetes officially
+
+
+## Apiserver spending too much cpu/memory in duplicate Endpoints serializations
+
+#### Issue
+
+Currently while serving watches, the apiserver is deep-copying deserialized endpoints objects from etcd and serializing it once for each kube-proxy watch it serves (which is 5k watches in our scale tests).
+
+#### Relevant issues/links
+
+- [#55779](https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-354452477)
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+- [#58050](https://github.com/kubernetes/kubernetes/issues/58050)
+
+#### Possible Solution(s)
+
+- Avoid those duplicate serializations if/when possible
+- Do better batching of EP updates in endpoints-controller: This is more like a bandaid, as the duplicate serializations would still happen, but we at least reduce the total no. of watch events (at the cost of increased latency of ep updates potentially). Also, this would not help too much with cases where pods of services are toggling between Ready and Failed states
```

Comment:
Personally I don't think it's a generic solution. It pretty much works only when you're createing/deleting a significant amount of service backends at once.
Since we probably need to solve other cases too, doing that may be unnecessary complication.

Reply:
Incorporated those points, PTAL.